norduser/utmp.go

```diff
@@ -0,0 +1,94 @@
+package norduser
+
+/*
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <utmp.h>
+
+const int ERROR_REALLOC = -1;
+const int ERROR_MALLOC_USERNAME = -2;
+
+void free_users_table(char*** users, int size) {
+  for (int index = 0; index < size; index++) {
+    if ((*users)[index] != NULL) {
+      free((*users)[index]);
+    }
+  }
+
+  if (*users != NULL) {
+    free(*users);
+  }
+}
+
+int get_utmp_user_processes(char*** users) {
+  int index = 0;
+  int size = 0;
+  setutent();
+
+  *users = NULL;
+
+  struct utmp* u;
+  while ((u = getutent()) != NULL) {
+    if (u->ut_type != USER_PROCESS) {
+      continue;
+    }
+
+    if (index == size) {
+      char** tmp;
+      tmp = realloc(*users, (size + 1) * sizeof(char*));
+      if (tmp == NULL) {
+        free_users_table(users, index);
+        return ERROR_REALLOC;
+      }
+      *users = tmp;
+
+      size++;
+    }
+
+    (*users)[index] = malloc(UT_NAMESIZE);
+    if ((*users)[index] == NULL) {
+      free_users_table(users, size);
+      return ERROR_MALLOC_USERNAME;
+    }
+
+    strcpy((*users)[index], u->ut_user);
```

Comment:
In `struct utmp` user string name is stored as string in `char ut_user[UT_NAMESIZE]`. So actual user name can be at most `UT_NAMESIZE - 1` because the last position needs to remain for null byte. So yeah, we can both use `strncpy(..., ..., UT_NAMESIZE)` and to be sure also write a null byte to `(*users)[index][UT_NAMESIZE-1]='\0';` There is no need to allocate `UT_NAMESIZE + 1`.

Reply:
but the documentations doesn't say it always has a null byte, just only if the name is smaller then the field size, `UT_NAMESIZE=32`. So if the user name has 32 chars, then `[UT_NAMESIZE-1]='\0';` would remove a char.

Reply:
> String fields are terminated by a null byte ('\0') if they are shorter than the size of the field.

You are right, I overlooked that fragment.